sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -384,22 +384,40 @@ object BooleanSimplification extends Rule[LogicalPlan] with PredicateHelper {
 /**
  * Simplifies binary comparisons with semantically-equal expressions:
  * 1) Replace '<=>' with 'true' literal.
- * 2) Replace '=', '<=', and '>=' with 'true' literal if both operands are non-nullable.
- * 3) Replace '<' and '>' with 'false' literal if both operands are non-nullable.
+ * 2) Replace '=', '<=', and '>=' with 'true' literal
+ *    if both operands are non-nullable or in constraints.
+ * 3) Replace '<' and '>' with 'false' literal if both operands are non-nullable or in constraints.
  */
-object SimplifyBinaryComparison extends Rule[LogicalPlan] with PredicateHelper {
+object SimplifyBinaryComparison
+  extends Rule[LogicalPlan] with PredicateHelper with ConstraintHelper {
+
+  private def canSimplifyWithConstraints(
+      plan: LogicalPlan, left: Expression, right: Expression): Boolean = {
+    def canSimplify(left: Expression, right: Expression): Boolean = {
+      !left.nullable && !right.nullable && left.semanticEquals(right)
+    }
+
+    if (SQLConf.get.constraintPropagationEnabled) {
+      canSimplify(left, right) || plan.constraints.exists {
```

Comment:
Can we limit this constrain check only in `Filter` plans?

Reply:
I think yes, this reply on `Filter` constraint which `InferFiltersFromConstraints` add.